lms/services/digest.py

```diff
@@ -138,6 +151,37 @@ def instructor_digest(self, h_userid):
                 # The user isn't an instructor in this course.
                 continue
 
+            course_assignments = []
+
+            for assignment_info in self.assignment_infos:
+                if (
+                    assignment_info.authority_provided_id
+                    != course_info.authority_provided_id
+                ):
+                    continue
+
+                assignment_learner_annotations = [
+                    annotation
+                    for annotation in unified_course.learner_annotations
+                    if annotation["assignment"]["tool_consumer_instance_guid"]
+                    == assignment_info.tool_consumer_instance_guid
+                    and annotation["assignment"]["resource_link_id"]
+                    == assignment_info.resource_link_id
+                ]
+
+                course_assignments.append(
+                    {
+                        "title": assignment_info.title,
+                        "num_annotations": len(assignment_learner_annotations),
+                        "annotators": list(
+                            set(
+                                annotation["author"]["userid"]
+                                for annotation in assignment_learner_annotations
+                            )
+                        ),
+                    }
+                )
```

Comment:
For each course we loop over `self.assignment_infos` and compute the list of assignments for that course (`course_assignments`) and add it to the course dict (which will be passed to the email template).

Note that we still separately count the total number of annotations and annotators in the course itself, as well as the numbers in each assignment in each course. This allows us to handle the case where we have an annotation but for whatever reason we don't have any assignment info in the DB for that annotation: the email template will first state the numbers of annotations and annotators in the course, and then for each known assignment in the course, and this will degrade nicely if we have annotations belonging to unknown assignments. Examples:

All a course's annotations belong to known assignments:

> **History of Jazz Music**
> 3 students made 8 annotations.
>
> * **Assignment One**
>   2 students made 5 annotations.
>
> * **Assignment Two**
>   A student made 2 annotations.
>
> * **Assignment Three**
>   A student made an annotation.

Some of a course's annotations belong to known assignments and some don't:

> **History of Jazz Music**
> 3 students made 8 annotations.
>
> * **Assignment One**
>   2 students made 5 annotations.
>
> * **Assignment Three**
>   A student made an annotation.

The number of annotations for the course is greater than the sum of the numbers for each assignment, because some of the course's annotations didn't belong to any known assignment. This may be a little misleading because teachers may not notice that the numbers don't add up or may not understand what it means if they don't. Still I think it's probably good enough? The alternative would be to list the unknown annotations under **Unknown Assignment(s)** (if there's more than one such annotation then we don't even know _how many_ different unknown assignments they belong to). I think the above approach is probably better.

When **all** of a course's annotations belong to unknown assignments I think this degrades quite nicely:

> **History of Jazz Music**
> 3 students made 8 annotations.